### ISSUE TITLE
Lz4 compression for unicode (non-english) data ?

### ISSUE THREAD

**Issue by PalanivelParthiban** [lz4/lz4#326]
When i try to compress the non-english data using lz4 compression, it returns the compressed size to be zero. Is there any way to make lz4 compression support unicode data ??

**This is the content of the file which i tried to compress.**

> The language samples below were generated using Google Translate.
> Welcome to Unicode and International language support for source materials in our product.
> [English]
> مرحبا بكم في ويونيكود الدعم الدولي لمواد اللغة مصدر في منتجاتنا.
> [Arabic]歡迎國際語言支持Unicode和源材料在我們的產品。
> [Chinese (Traditional)]
> Velkomin á Unicode og alþjóðleg tungumál styðja fyrir efni sem uppspretta í vörum okkar.
> [Icelandic]
> ようこそUnicodeと私たちの製品のソースの材料のための国際言語のサポート。
> [Japanese]

Thanks in advance.

**Comment by Cyan4973** [comment]
Compression may return `0` if there is not enough space in output buffer to contain the resulting compressed string.
Look at `LZ4_compressBound()`

**Comment by PalanivelParthiban** [comment]
@Cyan4973 

Even after trying LZ4_compressBound() in my code, I am getting the compressed size to be zero if I have  unicode data in my input buffer.
This is the code that I have used is given below
```

*out_data = new unsigned char [LZ4_compressBound(in_data_size)];
char *indata = (char*)in_data;
char *outdata = (char*)*out_data;
outsize = LZ4_compress_fast(indata,outdata,in_data_size,in_data_size,16);
```

**Comment by Cyan4973** [comment]
Try that : 
```
                size_t const out_data_capacity = LZ4_compressBound(in_data_size);
                *out_data = new unsigned char [out_data_capacity];
		char *indata = (char*)in_data;
		char *outdata = (char*)*out_data;
		outsize = LZ4_compress_fast(indata, outdata, in_data_size, out_data_capacity, 0);
```

**Comment by PalanivelParthiban** [comment]
Thanks for your input. But I have found another issue.

outsize = LZ4_compress_fast(indata, outdata, in_data_size, out_data_capacity, 0);

As you said this default compression is working for the test content that I have mentioned above. But for some other content it still returns zero value. It contatins some chinese and japanese contents. I couldn't mention it here because github comment box is not supporting those kind of format.

**Comment by Cyan4973** [comment]
Chinese and Japanese characters tend to be multi-bytes.
Details depend on the encoding format.
Consequently, there can be sometimes confusion between "number of characters" and "number of bytes". For LZ4 to work properly, it's necessary to use "number of bytes".
That could be something to check, or double-check.

Apart from that, I don't see any good reason why it would fail. `LZ4_compressBound()` is pretty well tested by now, and hasn't failed so far.

**Comment by PalanivelParthiban** [comment]
We tried compressing that Unicode data in various compression formats like 7z, winzip. It compresses those contents with very low compression ratio and takes more time than usual. For instance,

> Unicode content for about 800 MB is getting compressed to 790 MB in 2 to 3 minutes

This will actually degrades the overall speed of compression. So we have decided to write the particular block which contains the Unicode content ( LZ4_compress_fast() returns 0 ) as such in the compressed file. This block will be identified at the decompression through header contents.
